File: cpp/Mod_firstChild.cpp
#pragma once


Node firstChild(Tree t, Node n);
Node firstChild_debug(Tree t, Node n);

class firstChild_class {
    
    //VARIABLES
    Tree t;
    Tree t_spnew;
    Node n;
    Node n_spnew;
    Node res;
    Node res_spnew;
    bool debug_spv;
    
    //PREDICATES
    
    
    //OPERATIONS
    void o_init() {
        res_spnew = first(children(t, n));
        
        res = res_spnew;
        
        if(debug_spv) {
            cout << "o_init" << " ; "
            << "" << ";"  << "" << ";"  << res
            << endl;
        }
        
        return;
    }
    
public:
    firstChild_class(Tree constr_t, Node constr_n) : t(constr_t), n(constr_n)   {  }
    
    Node execute(bool debug) {
        debug_spv = debug;
        
        if(debug_spv) {
            cout << "firstChild" << ";"
            << "t"<< "; " << "n"<< "; " << "res"
            << endl;
            
            cout << "" << "; "
            << t << ";"  << n << ";"  << ""
            << endl;
        }
        
        o_init();
        return res;
    }
    
    
};

Node firstChild(Tree t, Node n) {
    firstChild_class obj(t, n);
    return obj.execute(false);
}


Node firstChild_debug(Tree t, Node n) {
    firstChild_class obj(t, n);
    return obj.execute(true);
}

File: hs/Mod_firstChild.hs
module Mod_firstChild ( firstChild ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types


trace' :: String -> a -> a
--trace' str x = (trace str x) 
trace' str x = x

data Data_firstChild  = Data_firstChild  {
  t :: Tree,
  n :: Node,
  res :: Node
}

--SIGNATURES
firstChild :: Tree -> Node -> Node
o_init :: Data_firstChild -> Node


--DEFINITIONS
firstChild t n = (o_init (Data_firstChild t n undefined))



o_init (Data_firstChild t n res) = (trace' "o_init" flow)
  where
    res' = (first (children t n))
    t' = t
    n' = n
    data' = (Data_firstChild t' n' res')
    flow =
      res'
